src/main/java/example/TriangleCount.java

```diff
@@ -0,0 +1,210 @@
+
+		Graph<String, HashSet<String>, NullValue> graphWithVertexNeighbors = Graph.fromDataSet(verticesWithLowerNeighbors,
+				edges, env);
+
+		// simulate the second superstep
+		// propagate each received "message" to neighbours with higher id
+		DataSet<Vertex<String, HashSet<String>>> verticesWithPropagatedValues =
+				graphWithVertexNeighbors.groupReduceOnNeighbors(new PropagateNeighborValues(), EdgeDirection.ALL);
+
+		Graph<String, HashSet<String>, NullValue> graphWithPropagatedVertexValues = Graph.fromDataSet(verticesWithPropagatedValues,
+				edges, env);
+
+		DataSet<Vertex<String, Integer>> verticesWithNumberOfTriangles = graphWithPropagatedVertexValues.getTriplets()
+				.map(new CountTriangles());
+
+		DataSet<Tuple1<Integer>> numberOfTriangles =  verticesWithNumberOfTriangles
+				.flatMap(new FlatMapFunction<Vertex<String, Integer>, Tuple1<Integer>>() {
+
+					@Override
+					public void flatMap(Vertex<String, Integer> vertex, Collector<Tuple1<Integer>> collector) throws Exception {
+						collector.collect(new Tuple1<Integer>(vertex.getValue()));
+					}
+				}).reduce(new ReduceFunction<Tuple1<Integer>>() {
+
+					@Override
+					public Tuple1<Integer> reduce(Tuple1<Integer> firstTuple, Tuple1<Integer> secondTuple) throws Exception {
+						return new Tuple1<Integer>(firstTuple.f0 + secondTuple.f0);
+					}
+				});
+
+		// emit result
+		if(fileOutput) {
+			numberOfTriangles.writeAsCsv(outputPath, "\n", ",");
+		} else {
+			numberOfTriangles.print();
+		}
+
+		env.execute("Executing Triangle Count");
+	}
+
+	@Override
+	public String getDescription() {
+		return "Triangle Count Example";
+	}
+
+	/**
+	 * Each vertex will have the set of neighbors with id lower than its own id as value.
+	 */
+	@SuppressWarnings("serial")
+	private static final class GatherLowerNeighbors implements
+			NeighborsFunctionWithVertexValue<String, NullValue, NullValue, Vertex<String, HashSet<String>>> {
+
+		@Override
+		public void iterateNeighbors(Vertex<String, NullValue> vertex,
+									 Iterable<Tuple2<Edge<String, NullValue>, Vertex<String, NullValue>>> neighbors,
+									 Collector<Vertex<String, HashSet<String>>> collector) throws Exception {
+
+			HashSet<String> neighborsHashSet = new HashSet<String>();
+			Tuple2<Edge<String, NullValue>, Vertex<String, NullValue>> next = null;
+			Iterator<Tuple2<Edge<String, NullValue>, Vertex<String, NullValue>>> neighborsIterator =
+					neighbors.iterator();
+
+			while (neighborsIterator.hasNext()) {
+				next = neighborsIterator.next();
+				if(Long.parseLong(next.f1.getId()) < Long.parseLong(vertex.getId())) {
+					neighborsHashSet.add(next.f1.getId());
+				}
+			}
+
+			collector.collect(new Vertex<String, HashSet<String>>(vertex.getId(), neighborsHashSet));
+		}
+	}
+
+	@SuppressWarnings("serial")
+	private static final class PropagateNeighborValues implements
+			NeighborsFunctionWithVertexValue<String, HashSet<String>, NullValue, Vertex<String, HashSet<String>>> {
```

Comment:
Same here. Each vertex should go through received messages, i.e. the elements inside its values and propagate each of the msgs to neighbors with higher ID, i.e. collect (neighborID, msg) pairs.